### PR TITLE
fix(invitations): better configurations selection + send explicit error when invitation is not sent

### LINKED ISSUE #1695: [Bug] Considérer les configurations des organisations de l'usager à l'invitation
Lorsqu'on invite un usager au niveau du territoire, on prend en compte pour son invitation une `configuration` qui vérifie les conditions suivantes: 
- elle est liée aux organisations de l'agent sur ce territoire
- elle est liée à la catégorie de motif depuis laquelle on invite l'usager

On peut voir comment on récupère a configuration dans le code [ici](https://github.com/betagouv/rdv-insertion/blob/d5efb6550271e9329eb8ab301eb6050b8652d5ba/app/services/invite_user.rb#L71) .

Le problème est que cette configuration n'est pas forcément liée à une organisation à laquelle appartient l'usager. Du coup, si l'usager est ajouté à une organisation A qui est configurée pour prendre des rdvs avec référents sur une catégorie de motif A', la configuration que l'on va récupérer lorsqu'on l'invite au niveau du territoire peut être liée à une organisation B et la catégorie de motif A', et cette configuration n'est pas forcément configurée pour prendre des rdvs avec référents.

Tout ça pour dire qu'il faut ajouter une condition lorsqu'on récupère les configurations disponibles [ici](https://github.com/betagouv/rdv-insertion/blob/d5efb6550271e9329eb8ab301eb6050b8652d5ba/app/services/invite_user.rb#L67): 
- Il faut récupérer les configurations liées aux organisations de l'agent (`@organisations`) **et** aux organisations de l'usager (`@user.organisations`) 

**Remarque** : 

Je pense que c'est possible qu'un agent puisse inviter un usager dans une catégorie qui n'est pas gérée par les organisations de l'usager (soit par la page index au niveau du territoire soit par API). Si c'est le cas il faudrait gérer ce cas d'erreur.

### LINKED ISSUE #1698: Signaler de manière explicite quand une invitation n'est pas envoyée
Lorsqu'une invitation a déjà été envoyée à un usager dans la journée dans un certain format, nous bloquons le renvoi de l'invitation, mais ce bloquage n'est pas signalé à l'agent : l'action semble s'être déroulée correctement. Il faudrait faire en sorte a minima que ce bloquage lui soit signalé au moment de l'envoi. On pourra ensuite réfléchir, en fonction des remontées, à lever ou à assouplir ce blocage dans le cadre de l'app (pour l'API ça a du sens)

### PR DESCRIPTION
**Pull request by qblanc**
closes #1695 
closes #1698 

Cette PR résout 2 faiblesses du service `InviteUser` : 
- les configurations prises en compte pour l'envoi de l'invitation doivent faire partie des organisations du `user` et pas seulement du `agent`
- lorsque nous bloquons l'envoi d'une invitation parce qu'une identique a déjà été envoyée le même jour, nous affichons une erreur à l'agent

### CODE REVIEW

File: app/services/invite_user.rb [hunk 1]
```diff
@@ -15,8 +15,7 @@ def call
     set_invitation_organisations
     set_invitation
     result.invitation = @invitation
-    return if invitation_already_sent_today? && !@invitation.format_postal?
-
+    check_if_invitation_should_be_sent
```
Comment (aminedhobb) [inline app/services/invite_user.rb:18]:
```suggestion
    check_if_invitation_should_be_sent!
```
pour être plus cohérent avec ce qu'on fait par ailleurs

File: app/services/invite_user.rb [hunk 2]
```diff
@@ -65,7 +70,7 @@ def motif_category
   end
 
   def all_configurations
-    @all_configurations ||= Configuration.where(organisation: @organisations).to_a
+    @all_configurations ||= Configuration.where(organisation: @organisations & @user.organisations).to_a
   end
 
   def set_current_configuration
```
Comment (aminedhobb) [inline app/services/invite_user.rb:76]: Mais du coup si on est dans le edge case que j'évoque dans le ticket, `@current_configuration` sera `nil` et on aura une erreur 500. 
Il faudrait faire une vérif ici je pense. On fait déjà une vérif [ici](https://github.com/betagouv/rdv-insertion/blob/d5efb6550271e9329eb8ab301eb6050b8652d5ba/app/services/invitations/validate.rb#L47) mais ça intervient plus/trop tard

**Review by aminedhobb — COMMENTED** [review]
Je t'ai laissé un commentaire sur le edge case 👍

File: app/services/invite_user.rb (hunk 2 above)
Comment (qblanc) [reply app/services/invite_user.rb:76]: Ok, j'ai fait en sorte que ce edge case ne soit plus possible dans le nouveau commit

**Review by aminedhobb — COMMENTED** [review]
Merci @qblanc 🙏 !
C'est top, je me demande du coup si on devrait pas utiliser ce service (ou un nouveau service `RdvContexts::FindOrCreate`) dans le service `InviteUser` au niveau de la méthode `find_or_create_rdv_contexts`.
(Si on fait ça, ça pose la question de la pertinence de garder le check dans `Invitations::Validate`).

**Comment by aminedhobb** [comment]
> Merci @qblanc 🙏 ! C'est top, je me demande du coup si on devrait pas utiliser ce service (ou un nouveau service `RdvContexts::FindOrCreate`) dans le service `InviteUser` au niveau de la méthode `find_or_create_rdv_contexts`. (Si on fait ça, ça pose la question de la pertinence de garder le check dans `Invitations::Validate`).

Si tu veux on peut séparer ça en deux PRs et garder cette PR pour le premier fix et faire une autre pour gérer les cas des contextes / organisations

File: app/jobs/inbound_webhooks/rdv_solidarites/process_rdv_job.rb [hunk 3]
```diff
@@ -232,9 +232,7 @@ def invalidate_related_invitations
       def rdv_contexts
         @rdv_contexts ||=
           @users.map do |user|
-            RdvContext.with_advisory_lock "setting_rdv_context_for_user_#{user.id}" do
-              RdvContext.find_or_create_by!(user: user, motif_category: motif_category)
-            end
+            RdvContexts::FindOrCreate.call(user:, motif_category:).rdv_context
```
Comment (aminedhobb) [inline app/jobs/inbound_webhooks/rdv_solidarites/process_rdv_job.rb:235]: Du coup le cas d'erreur n'est pas géré ici donc ça aura un comportement inattendu quand le service va fail

File: app/services/invite_user.rb [hunk 4]
```diff
@@ -50,22 +49,27 @@ def set_invitation_organisations
...
   def invitation_already_sent_today?
     @rdv_context.invitations.where(format: @invitation.format).where("sent_at > ?", 24.hours.ago).present?
   end
 
   def find_or_create_rdv_context
-    RdvContext.with_advisory_lock "setting_rdv_context_for_user_#{@user.id}" do
-      @rdv_context = RdvContext.find_or_create_by!(motif_category: @current_configuration.motif_category, user: @user)
-    end
+    @rdv_context =
+      RdvContexts::FindOrCreate.call(user: @user, motif_category: @current_configuration.motif_category).rdv_context
```
Comment (aminedhobb) [inline app/services/invite_user.rb:64]: pareil ici le cas d'erreur n'est pas géré

File: app/services/rdv_contexts/find_or_create.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+      return if rdv_context_already_exists?
+
+      create_rdv_context!
+    end
+
+    def create_rdv_context!
+      RdvContext.transaction do
+        @rdv_context = RdvContext.new(user: @user, motif_category: @motif_category)
+        check_if_user_has_an_organisation_for_this_motif_cateogry!
+        authorize @rdv_context, :create?
```
Comment (aminedhobb) [inline app/services/rdv_contexts/find_or_create.rb:29]: On met les autorisations dans le controller en général, notamment parce que si la personne n'est pas autorisé ça va raise une `Pundit::NotAuthorizedError` qui est rescued au niveau des controllers mais pas ici

**Review by aminedhobb — CHANGES_REQUESTED** [review]
Merci @qblanc, je pense qu'il y a quelques choses à changer sur cette dernière implémentation.
Mais comme on a pas encore shippé le bug initial qui est impactant et qu'on est en train de gérer des edge cases, je serais partant qu'on revienne sur sur [ton dernier commit](https://github.com/betagouv/rdv-insertion/pull/1699/commits/e405a72940636ffec7e15b53b2b1c08fcd1e8678) (en ajoutant peut-être un test au service) et qu'on gère ces cas-là plus tard (j'écrirai un ticket). C'est bon pour toi ?

**Comment by qblanc** [comment]
ha ha oui, moi ça me va de revenir au commit précédent, c'est toi qui me demandait de rajouter une nouvelle brique à chaque fois :)

**Review by aminedhobb — DISMISSED** [review]
Merci @qblanc 🙏 !!
